### PR TITLE
num-audit: extension

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

Extension of #454 

## Solution

* gas related fields to `u128`
* `nonce` and `number` to `u64` in `Header` from `B64` and `U256` respectively

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [x] Breaking changes

### CODE REVIEW

File: crates/eips/src/eip1559/basefee.rs [hunk 1]
```diff
@@ -7,17 +7,17 @@ use crate::eip1559::constants::{
...
+    pub elasticity_multiplier: u128,
 }
 
 impl BaseFeeParams {
     /// Get the base fee parameters for Ethereum mainnet
     pub const fn ethereum() -> BaseFeeParams {
         BaseFeeParams {
-            max_change_denominator: DEFAULT_BASE_FEE_MAX_CHANGE_DENOMINATOR,
-            elasticity_multiplier: DEFAULT_ELASTICITY_MULTIPLIER,
+            max_change_denominator: DEFAULT_BASE_FEE_MAX_CHANGE_DENOMINATOR as u128,
```
Comment (prestwich) [inline crates/eips/src/eip1559/basefee.rs:19]: should we change the constant types?

File: crates/rpc-types/src/eth/block.rs [hunk 2]
```diff
@@ -75,11 +75,11 @@ pub struct Header {
...
-    pub number: Option<U256>,
+    pub number: Option<u64>,
     /// Gas Limit
-    pub gas_limit: U256,
+    pub gas_limit: u128,
     /// Gas Used
-    pub gas_used: U256,
+    pub gas_used: u128,
     /// Timestamp
     pub timestamp: U256,
```
Comment (prestwich) [inline crates/rpc-types/src/eth/block.rs:84]: maybe followup work, but should `timestamp` be a `u64` cc @mattsse?

File: crates/provider/src/heart.rs [hunk 3]
```diff
@@ -445,12 +445,12 @@ impl<S> Heartbeat<S> {
             // Otherwise add it to the waiting list.
             debug!(tx=%watcher.config.tx_hash, %block_height, confirmations, "adding to waiting list");
             self.waiting_confs
-                .entry(*block_height + U256::from(confirmations) - U256::from(1))
+                .entry(U256::from(*block_height + confirmations - 1))
```
Comment (prestwich) [inline crates/provider/src/heart.rs:448]: we could change the type of `waiting_confs` to `BTreeMap<u64, Vec<TxWatcher>>,`

**Comment by klkvr** [comment]
`rpc_types::TransactionReceipt` also has several `u64` gas price fields

File: crates/rpc-types/src/eth/block.rs (hunk 2 above)
Comment (yash-atreya) [reply crates/rpc-types/src/eth/block.rs:84]: `timestamp` is `u64` in certain places. I'm in favour

**Comment by klkvr** [comment]
I am experiencing https://github.com/serde-rs/json/issues/625 on latest alloy while deserializing `WithOtherFields<TransactionRequest>`, so I think we should ensure that all `u128` fields have a `with = u128_hex_or_decimal`

File: crates/eips/src/eip1559/basefee.rs (hunk 1 above)
Comment (yash-atreya) [reply crates/eips/src/eip1559/basefee.rs:19]: Ideally yes, but I think if users would like to use constants in their applications; It's smoother for them to use `as u128` and expand to a large type than use `try_from().into()` to shrink to a smaller.

Comment (yash-atreya) [reply crates/eips/src/eip1559/basefee.rs:19]: Moreover, imo, its better to expand the type in places where its absolutely needed and keep it as is where isn't required.

File: crates/rpc-types/src/eth/block.rs (hunk 2 above)
Comment (mattsse) [reply crates/rpc-types/src/eth/block.rs:84]: sg!

File: crates/consensus/src/header.rs [hunk 4]
```diff
@@ -397,8 +397,8 @@ impl Decodable for Header {
             logs_bloom: Decodable::decode(buf)?,
             difficulty: Decodable::decode(buf)?,
             number: U256::decode(buf)?.to::<u64>(),
-            gas_limit: U256::decode(buf)?.to::<u64>(),
-            gas_used: U256::decode(buf)?.to::<u64>(),
+            gas_limit: U256::decode(buf)?.to::<u128>(),
+            gas_used: U256::decode(buf)?.to::<u128>(),
```
Comment (mattsse) [inline crates/consensus/src/header.rs:401]: I think we should either use u128::decode or use try_into because to:: is not safe

File: crates/eips/src/eip1559/basefee.rs
Comment (mattsse) [inline crates/eips/src/eip1559/basefee.rs:1]: after thinking about these a bit, I think these make sense, if we change fees/gas used to u128 then all of these also make sense

File: crates/rpc-types/src/eth/block.rs [hunk 5]
```diff
@@ -75,11 +75,11 @@ pub struct Header {
...
     pub difficulty: U256,
     /// Block number
-    pub number: Option<U256>,
+    pub number: Option<u64>,
     /// Gas Limit
-    pub gas_limit: U256,
+    pub gas_limit: u128,
     /// Gas Used
-    pub gas_used: U256,
+    pub gas_used: u128,
```
Comment (mattsse) [inline crates/rpc-types/src/eth/block.rs:82]: these need serde attributes

File: crates/rpc-types/src/eth/block.rs [hunk 6]
```diff
@@ -101,19 +101,19 @@ pub struct Header {
...
     #[serde(skip_serializing_if = "Option::is_none")]
     pub withdrawals_root: Option<B256>,
     /// Blob gas used
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub blob_gas_used: Option<U64>,
+    pub blob_gas_used: Option<u128>,
     /// Excess blob gas
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub excess_blob_gas: Option<U64>,
+    pub excess_blob_gas: Option<u128>,
```
Comment (mattsse) [inline crates/rpc-types/src/eth/block.rs:116]: same

**Review by mattsse — CHANGES_REQUESTED** [review]
very supportive
unifying all of them is necessary, just need to ensure we keep hex support

File: crates/serde/src/num.rs [hunk 7]
```diff
@@ -243,7 +243,6 @@ where
 {
     NumberOrHexU256::deserialize(deserializer)?.try_into_u256()
 }
-
```
Comment (prestwich) [inline crates/serde/src/num.rs:246]: nit:
was this change intentional?

**Review by prestwich — APPROVED** [review]

File: crates/serde/src/num.rs (hunk 7 above)
Comment (yash-atreya) [reply crates/serde/src/num.rs]: That was a mistake

File: crates/rpc-types/src/eth/block.rs [hunk 8]
```diff
@@ -75,10 +75,13 @@ pub struct Header {
     /// Difficulty
     pub difficulty: U256,
     /// Block number
+    #[serde(default, with = "alloy_serde::num::u64_hex_opt")]
     pub number: Option<u64>,
     /// Gas Limit
+    #[serde(default, with = "alloy_serde::num::u128_hex_or_decimal")]
     pub gas_limit: u128,
     /// Gas Used
+    #[serde(default, with = "alloy_serde::num::u128_hex_or_decimal")]
```
Comment (mattsse) [inline crates/rpc-types/src/eth/block.rs]: this should just be hex

File: crates/rpc-types/src/eth/block.rs [hunk 9]
```diff
@@ -101,18 +104,31 @@ pub struct Header {
...
+        with = "alloy_serde::num::u128_hex_or_decimal_opt"
+    )]
     pub blob_gas_used: Option<u128>,
     /// Excess blob gas
-    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        with = "alloy_serde::num::u128_hex_or_decimal_opt"
+    )]
```
Comment (mattsse) [inline crates/rpc-types/src/eth/block.rs]: also just hex

**Review by mattsse — CHANGES_REQUESTED** [review]
last pedantic serde nits

File: crates/eips/src/eip1559/basefee.rs [hunk 10]
```diff
@@ -7,17 +7,17 @@ use crate::eip1559::constants::{
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 pub struct BaseFeeParams {
     /// The base_fee_max_change_denominator from EIP-1559
-    pub max_change_denominator: u64,
+    pub max_change_denominator: u128,
     /// The elasticity multiplier from EIP-1559
-    pub elasticity_multiplier: u64,
+    pub elasticity_multiplier: u128,
```
Comment (klkvr) [inline crates/eips/src/eip1559/basefee.rs:12]: This needs serde attr

File: crates/genesis/src/lib.rs [hunk 11]
```diff
@@ -64,14 +66,14 @@ pub struct Genesis {
...
     /// The genesis header excess blob gas
-    #[serde(skip_serializing_if = "Option::is_none", with = "u64_hex_or_decimal_opt")]
-    pub excess_blob_gas: Option<u64>,
+    #[serde(skip_serializing_if = "Option::is_none", with = "u128_hex_or_decimal_opt")]
+    pub excess_blob_gas: Option<u128>,
     /// The genesis header blob gas used
-    #[serde(skip_serializing_if = "Option::is_none", with = "u64_hex_or_decimal_opt")]
-    pub blob_gas_used: Option<u64>,
+    #[serde(skip_serializing_if = "Option::is_none", with = "u128_hex_or_decimal_opt")]
+    pub blob_gas_used: Option<u128>,
```
Comment (klkvr) [inline crates/genesis/src/lib.rs:76]: Those need `#[serde(default)]` ref https://github.com/alloy-rs/alloy/pull/370

Comment (mattsse) [reply crates/genesis/src/lib.rs:76]: good catch,
always forget that missing key is not handled by option

File: crates/eips/src/eip1559/basefee.rs (hunk 10 above)
Comment (mattsse) [reply crates/eips/src/eip1559/basefee.rs:12]: it's not part of rpc, but we should do hex here yes

File: crates/rpc-types/src/eth/transaction/receipt.rs [hunk 12]
```diff
@@ -28,21 +28,29 @@ pub struct TransactionReceipt<T = ReceiptEnvelope<Log>> {
     #[serde(with = "alloy_serde::u64_hex_opt")]
     pub block_number: Option<u64>,
     /// Gas used by this transaction alone.
-    #[serde(with = "alloy_serde::u64_hex_opt")]
-    pub gas_used: Option<u64>,
+    #[serde(with = "alloy_serde::u128_hex_or_decimal_opt")]
+    pub gas_used: Option<u128>,
```
Comment (klkvr) [inline crates/rpc-types/src/eth/transaction/receipt.rs:32]: needs `default` and `skip_serializing_if`

aren't those mandatory though?

Comment (klkvr) [reply crates/rpc-types/src/eth/transaction/receipt.rs:32]: image

File: crates/rpc-types/src/eth/block.rs [hunk 13]
```diff
@@ -101,19 +105,32 @@ pub struct Header {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub mix_hash: Option<B256>,
     /// Nonce
-    pub nonce: Option<B64>,
+    #[serde(default, with = "alloy_serde::num::u64_hex_opt")]
+    pub nonce: Option<u64>,
```
Comment (klkvr) [inline crates/rpc-types/src/eth/block.rs:109]: `skip_serializing_if`

File: crates/rpc-types/src/eth/block.rs (hunk 8 above)
Comment (yash-atreya) [reply crates/rpc-types/src/eth/block.rs]: Going to address this in a separate PR for https://github.com/alloy-rs/alloy/issues/465

File: crates/rpc-types/src/eth/block.rs (hunk 9 above)
Comment (yash-atreya) [reply crates/rpc-types/src/eth/block.rs]: Going to address this in a separate PR for https://github.com/alloy-rs/alloy/issues/465

File: crates/rpc-types/src/eth/block.rs [hunk 14]
```diff
@@ -75,13 +75,17 @@ pub struct Header {
     /// Difficulty
     pub difficulty: U256,
     /// Block number
-    pub number: Option<U256>,
+    #[serde(default, with = "alloy_serde::num::u64_hex_opt")]
+    pub number: Option<u64>,
```
Comment (klkvr) [inline crates/rpc-types/src/eth/block.rs:79]: `skip_serializing_if` and feels like this is mandatory too?

File: crates/rpc-types/src/eth/transaction/request.rs [hunk 15]
```diff
@@ -14,18 +14,23 @@ pub struct TransactionRequest {
...
+    #[serde(default, with = "alloy_serde::num::u128_hex_or_decimal_opt")]
     pub gas_price: Option<u128>,
     /// The max base fee per gas the sender is willing to pay.
-    #[serde(default)]
+    #[serde(default, with = "alloy_serde::num::u128_hex_or_decimal_opt")]
     pub max_fee_per_gas: Option<u128>,
     /// The max priority fee per gas the sender is willing to pay, also called the miner tip.
-    #[serde(default)]
+    #[serde(default, with = "alloy_serde::num::u128_hex_or_decimal_opt")]
     pub max_priority_fee_per_gas: Option<u128>,
```
Comment (klkvr) [inline crates/rpc-types/src/eth/transaction/request.rs:24]: `skip_serializing_if`

Comment (prestwich) [reply crates/rpc-types/src/eth/transaction/request.rs:24]: isnt this handled by the serde_with?

File: crates/rpc-types/src/eth/block.rs (hunk 14 above)
Comment (mattsse) [reply crates/rpc-types/src/eth/block.rs:79]: number is acutally required, yes

File: crates/rpc-types/src/eth/transaction/receipt.rs (hunk 12 above)
Comment (mattsse) [reply crates/rpc-types/src/eth/transaction/receipt.rs:32]: gas used is required:

```
// Receipt represents the results of a transaction.
type Receipt struct {
	// Consensus fields: These fields are defined by the Yellow Paper
	Type              uint8  `json:"type,omitempty"`
	PostState         []byte `json:"root"`
	Status            uint64 `json:"status"`
	CumulativeGasUsed uint64 `json:"cumulativeGasUsed" gencodec:"required"`
	Bloom             Bloom  `json:"logsBloom"         gencodec:"required"`
	Logs              []*Log `json:"logs"              gencodec:"required"`

	// Implementation fields: These fields are added by geth when processing a transaction.
	TxHash            common.Hash    `json:"transactionHash" gencodec:"required"`
	ContractAddress   common.Address `json:"contractAddress"`
	GasUsed           uint64         `json:"gasUsed" gencodec:"required"`
	EffectiveGasPrice *big.Int       `json:"effectiveGasPrice"` // required, but tag omitted for backwards compatibility
	BlobGasUsed       uint64         `json:"blobGasUsed,omitempty"`
	BlobGasPrice      *big.Int       `json:"blobGasPrice,omitempty"`

	// Inclusion information: These fields provide information about the inclusion of the
	// transaction corresponding to this receipt.
	BlockHash        common.Hash `json:"blockHash,omitempty"`
	BlockNumber      *big.Int    `json:"blockNumber,omitempty"`
	TransactionIndex uint        `json:"transactionIndex"`
}
```

File: crates/rpc-types/src/eth/transaction/request.rs (hunk 15 above)
Comment (klkvr) [reply crates/rpc-types/src/eth/transaction/request.rs:24]: it will be serialized as null by default

File: crates/rpc-types/src/eth/block.rs (hunk 13 above)
Comment (yash-atreya) [reply crates/rpc-types/src/eth/block.rs:109]: I think even nonce is required

Comment (mattsse) [reply crates/rpc-types/src/eth/block.rs:113]: mix_hash and nonce are options because they are not supported by newer chains that otherwise are compatible with this object.
for convenience I'd like to keep them optional

File: crates/rpc-types/src/eth/transaction/receipt.rs (hunk 12 above)
Comment (mattsse) [reply crates/rpc-types/src/eth/transaction/receipt.rs:32]: @yash-atreya

File: crates/rpc-types/src/eth/transaction/receipt.rs [hunk 16]
```diff
@@ -28,21 +28,29 @@ pub struct TransactionReceipt<T = ReceiptEnvelope<Log>> {
...
     /// The price paid post-execution by the transaction (i.e. base fee + priority fee). Both
     /// fields in 1559-style transactions are maximums (max fee + max priority fee), the amount
     /// that's actually paid by users can only be determined post-execution
-    #[serde(with = "alloy_serde::u64_hex")]
-    pub effective_gas_price: u64,
+    #[serde(with = "alloy_serde::u128_hex_or_decimal")]
```
Comment (mattsse) [inline crates/rpc-types/src/eth/transaction/receipt.rs:36]: only hex for all rpc fields

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc-types/src/eth/block.rs (hunk 14 above)
Comment (mattsse) [reply crates/rpc-types/src/eth/block.rs:79]: there are no clear rules for the pending block,
until we figure it out, let's keep them optional

File: crates/rpc-types/src/eth/transaction/receipt.rs (hunk 16 above)
Comment (yash-atreya) [reply crates/rpc-types/src/eth/transaction/receipt.rs:36]: Yes, I believe there should be a separate PR for #465 as it's likely to break several tests, and I also wanted to unblock https://github.com/foundry-rs/foundry/pull/7106 quickly. I'll implement it immediately after this.

**Comment by prestwich** [comment]
#465 actually somewhat superseded by the newer issue wrt serde helpers redundant code

**Comment by yash-atreya** [comment]
> #465 actually somewhat superseded by the newer issue wrt serde helpers redundant code

Yes, the PR that addresses #479 will close #465

**Comment by prestwich** [comment]
> > #465 actually somewhat superseded by the newer issue wrt serde helpers redundant code
> 
> 
> 
> Yes, the PR that addresses #479 will close #465

not exactly because moving to `via_ruint` canonizes the tolerant behavior, where #465 would require significant changes to reject the invalid-but-tolerated objects

Given how much surface area its add, my current preference is to leave everything tolerant

**Comment by mattsse** [comment]
> Given how much surface area its add, my current preference is to leave everything tolerant

hmm, for response types, being tolerant is actually beneficial because the user doesn't really care how the server formats it

**Review by mattsse — APPROVED** [review]
per above, 
also, supporting decimals on deserialize is okay and perhaps even preferable, in case servers don't follow QUANTITY
